### PR TITLE
Graph break on step closure in optimizer

### LINKED ISSUE #116494: Dynamo'd optimizer does not handle closure correctly
### 🐛 Describe the bug

My suspicion is that we do not re-enable grad mode for the closure when calling it from within step. Due to this, there is a disparity between eager and dynamo for optimizers with closures that depend on grads getting modified, which...according to https://github.com/pytorch/pytorch/issues/38006, account for all cases where closures are used.

Disclaimer: There is an ongoing issue (https://github.com/pytorch/pytorch/issues/38006) that requires looking into where the example I provide below is potentially abusing the closure (I modify the .grad field of the params within the closure), which the docs say should not happen, but empirical evidence points 100% otherwise. Regardless, we probably want to either ban closures (and error when we notice one) or actually call them correctly.

### Error logs

```
(pytorch-3.10) [janeyx@devgpu023.odn1 ~/local/pytorch (d186986c)]$ python playground2.py 
before compiled loop: param=tensor([[0.6176, 0.6603, 0.8097],
        [0.8297, 0.5115, 0.0299]], device='cuda:0', requires_grad=True)
after compiled loop: param=tensor([[0.6176, 0.6603, 0.8097],
        [0.8297, 0.5115, 0.0299]], device='cuda:0', requires_grad=True)
before eager loop: param_c=tensor([[0.6176, 0.6603, 0.8097],
        [0.8297, 0.5115, 0.0299]], device='cuda:0', requires_grad=True)
after eager loop: param_c=tensor([[0.6166, 0.6593, 0.8087],
        [0.8287, 0.5105, 0.0289]], device='cuda:0', requires_grad=True)
```

### Minified repro

```
import torch

param = torch.rand(2, 3, dtype=torch.float32, device='cuda', requires_grad=True)
param_c = param.clone().detach().requires_grad_(True)

def closure():
    param.grad = torch.ones_like(param) * 2
    return param.grad

def closure_c():
    param_c.grad = torch.ones_like(param_c) * 2
    return param_c.grad

optimizer = torch.optim.AdamW([param])
optimizer_c = torch.optim.AdamW([param_c])

def loop(opt, c):
    opt.step(c)

compiled_loop = torch._dynamo.optimize("eager")(loop)

print(f"before compiled loop: {param=}")
compiled_loop(optimizer, closure)
print(f"after compiled loop: {param=}")

print(f"before eager loop: {param_c=}")
loop(optimizer_c, closure_c)
print(f"after eager loop: {param_c=}")
```

### Versions

on main

cc @ezyang @gchanan @zou3519 @kadeng @vincentqb @jbschlosser @albanD @crcrpar @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @mlazos

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/pytorch/issues/116494

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_dynamo/symbolic_convert.py [hunk 1]
```diff
@@ -2121,6 +2124,19 @@ def __init__(
...
+        if (
+            "closure" in self.symbolic_locals
+            and not isinstance(self.symbolic_locals["closure"], ConstantVariable)
+            and "self" in self.symbolic_locals
+            and isinstance(self.symbolic_locals["self"], OptimizerVariable)
+            and isinstance(self.symbolic_locals["self"].value, torch.optim.Optimizer)
+            and self.code_options["co_name"] == "step"
+        ):
+            unimplemented("Optimizer step with closure not supported.")
+
```
Comment (yanboliang) [inline torch/_dynamo/symbolic_convert.py:2139]: I'm curious if we can merge this and the changes at optimizer.py into the same place, it seems they do same thing. Can we define a rule at ```trace_rule.lookup_inner``` like if function is optimizer.step and first arg is closure, then graph break?

Comment (mlazos) [reply torch/_dynamo/symbolic_convert.py:2139]: yeah that sounds awesome! didn't know it was possible. I was pretty annoyed to have to include a case for "if we're tracing" and "if we are compiling that function"

**Review by yanboliang — APPROVED** [review]

**Pull request closed without merging**